packages/snip20/src/query.rs

```diff
@@ -0,0 +1,315 @@
+pub struct Balance {
+    pub amount: Uint128,
+}
+/// Transaction data
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct Tx {
+    pub id: u64,
+    pub from: HumanAddr,
+    pub sender: HumanAddr,
+    pub receiver: HumanAddr,
+    pub coins: Coin,
+}
+/// TransferHistory response
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct TransferHistory {
+    pub txs: Vec<Tx>,
+}
+/// Minters response
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct Minters {
+    pub minters: Vec<HumanAddr>,
+}
+
+/// SNIP20 queries
+#[derive(Serialize)]
+#[serde(rename_all = "snake_case")]
+pub enum Snip20QueryMsg<'a> {
+    TokenInfo {},
+    ExchangeRate {},
+    Allowance {
+        owner: &'a HumanAddr,
+        spender: &'a HumanAddr,
+        key: &'a str,
+    },
+    Balance {
+        address: &'a HumanAddr,
+        key: &'a str,
+    },
+    TransferHistory {
+        address: &'a HumanAddr,
+        key: &'a str,
+        page: Option<u32>,
+        page_size: u32,
+    },
+    Minters {},
+}
+
+impl<'a> fmt::Display for Snip20QueryMsg<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Snip20QueryMsg::TokenInfo { .. } => write!(f, "TokenInfo"),
+            Snip20QueryMsg::ExchangeRate { .. } => write!(f, "ExchangeRate"),
+            Snip20QueryMsg::Allowance { .. } => write!(f, "Allowance"),
+            Snip20QueryMsg::Balance { .. } => write!(f, "Balance"),
+            Snip20QueryMsg::TransferHistory { .. } => write!(f, "TransferHistory"),
+            Snip20QueryMsg::Minters { .. } => write!(f, "Minters"),
+        }
+    }
+}
+
+impl<'a> Snip20QueryMsg<'a> {
+    /// Returns a StdResult<T>, where T is the "Response" type that wraps the query answer
+    ///
+    /// # Arguments
+    ///
+    /// * `deps` - reference to Extern that holds all the external contract dependencies
+    /// * `block_size` - pad message to blocks of this size
+    /// * `callback_code_hash` - string slice holding the code hash of contract being queried
+    /// * `contract_addr` - reference to address of contract being queries
+    pub fn query<S: Storage, A: Api, Q: Querier, T: DeserializeOwned>(
+        &'a self,
+        deps: &'a Extern<S, A, Q>,
+        block_size: usize,
+        callback_code_hash: &'a str,
+        contract_addr: &'a HumanAddr,
+    ) -> StdResult<T> {
+        let pad_block_size: usize;
+        // can not have block size of 0
+        if block_size == 0 {
+            pad_block_size = 1;
+        } else {
+            pad_block_size = block_size;
+        }
+        let mut msg = to_binary(self)?;
+        space_pad(&mut msg.0, pad_block_size);
+        deps.querier
+            .query(&QueryRequest::Wasm(WasmQuery::Smart {
+                contract_addr: contract_addr.clone(),
+                callback_code_hash: callback_code_hash.to_string(),
+                msg,
+            }))
```

Comment:
in cases like this, where you know the function needs an owned instance of an object, it's better to require an owned instance of it in the function parameters (in this case you only required `&HumanAddr` and `&str`). That way either the user is required to make a clone themselves, which exposes the extra cost, or the user avoids the clone, if they don't need the owned object for anything else, and they can just pass it in.

Reply:
@reuvenpo OK.  Honestly I still don't really have a feel for how the rust compiler handles passing by value.  I was thinking that passing by ref would avoid having to copy the pointer/length/capacity portion of the String when they call the name-specific function and avoid it again when that function calls the `query` function.  But from your comment it sounds like if the users knows they won't need the Strings again, since they just avoid doing the clone (which is probably more costly than the multiple pointer/length/capacity copies anyway), it saves costs in those cases.

Reply:
I think reading this will help you understand these things better. I think it'll explain things much better and more thouroughly than I can, but if you have mose specific questions, feel free to PM. 
https://doc.rust-lang.org/book/ch04-01-what-is-ownership.html

The one thing i'll add is that what you wrote is correct, plus cloning a string also involves heap allocation and full content copying, which is significantly more expensive than copying three fields on the stack.